teps/0015-pause-pipeline.md

```diff
@@ -0,0 +1,105 @@
+ignored by the Tekton reconciler. These paused resources can be used by platform
+implementers to control PipelineRun and TaskRun resources, for example, in cases
+where the cluster is under heavy load and not ready to process a PipelineRun.
+
+PipelineRuns and TaskRuns can be created paused so that they are not started at
+all or, in the case of a PipelineRun, the pause setting can be used to prevent
+further TaskRuns from being created. As it is not possible to "pause" a
+Kubernetes pod, the pause setting would not affect TaskRuns after they are
+started.
+
+## Motivation
+
+The primary motivation for this is to support operators who need to control the
+start up of PipelineRuns and TaskRuns in clusters that are under heavy load.
+By allowing PipelineRuns to be created without actually starting them, one
+could implement a custom controller to implement any PipelineRun scheduling
+policy that is transparent to the PipelineRun controller and PipelineRun creator
+(e.g., Tekton triggers).
+
+With this feature, platforms can control when PipelineRuns and TaskRuns are
+started without introducing a bespoke system for storing PipelineRuns -
+potentially adding new failure scenarios or UI complexity.
+
+### Goals
+
+* Provide a pause mechanism for PipelineRuns and TaskRuns that operators can use
+  to control PipelineRuns and TaskRuns.
+
+### Non-Goals
+
+* Support pausing of a running TaskRun as that would requiring pausing the
+  Kubernetes pods themselves.
+
+## Requirements
+
+* Provide a mechanism to operators that can be used to pause a PipelineRun or
+  TaskRun.
+* PipelineRuns or TaskRuns that are created with the pause setting enabled are
+  not started until the setting is toggled off.
+* PipelineRuns that have the pause setting enabled after they are started will
+  not create any new TaskRuns until the setting is toggled off.
+
+## Proposal
+
+Both the PipelineRunSpec and TaskRunSpec support a field called `Status`. The
+`Status` field is either a `PipelineRunSpecStatus` or a `TaskRunSpecStatus`
+resource, respectively. Currently these fields only support a single setting,
+`PipelineRunCancelled` or `TaskRunCancelled`. This field will be expanded to
+support an additional setting, `PipelineRunPaused` for `PipelineRuns` and
+`TaskRunPaused` for `TaskRuns`.
+
+When the `Status` field is set to `PipelineRunPaused` or `TaskRunPaused`, the
+Tekton reconciler will ignore updates to the `PipelineRun` or `TaskRun`
+resources.
+
+To "un-pause" a resource, the operator should set the `Status` field to empty,
+the Tekton controller will see the change and proceed with the Pipeline or Task
+as normal.
+
+
+### User Stories
+
+#### Platform Implementor
+
+As a platform implementor, I want to be able to control when a Pipeline or Task
+is started without introducing bespoke machinery.
+
+## Alternatives
+
+* Instead of adding this setting directly to the PipelineRun or TaskRun object,
```

Comment:
Spit balling here. This might be an opportunity make to make custom interceptor with something like rabbitMQ. So, when a event is made you can broker it yourself. 🤔

Reply:
```
* PipelineRuns that are not ready to be started could be stored in a
  non-Kubernetes-native queue, such as Redis, and only submitted once they are
  ready to run. This adds complexity in that a new data store must be run and
  it is not easy to provide users visibility into the queue.
```

Reply:
Sorry, I should clarify. Tekton triggers can take events and map them to pipelineruns. There a list of (interceptors)[https://github.com/tektoncd/triggers/tree/master/pkg/interceptors] that can manipulate payloads and do extra processing. I agree this approach isn't the best solution. I just think it would be easier to schedule the events rather than pipeline/tasks themselves.

Reply:
I was just wondering if this is an event issue. 😅

Reply:
This actually just a variation on this comment. > This is actually possible today, I think. You could have a mutating admission webhook force new PipelineRuns into the Succeeded=Unknown state (so Tekton doesn't try to start them immediately) and a reconciling controller with permission to update PipelineRun statuses to push them into a running state, at which point Tekton's normal reconciliation would take over and start any TaskRuns.